Mathlib/CategoryTheory/Grothendieck.lean

```diff
@@ -149,6 +153,108 @@ def forget : Grothendieck F ⥤ C where
+  obj X := {
+    base := X.base
+    fiber := (α.app X.base).obj X.fiber
+  }
+  map {X Y} f := {
+    base := f.base
+    fiber := (eqToHom (α.naturality f.base).symm).app X.fiber ≫ (α.app Y.base).map f.fiber
+  }
+  map_id := by
+    intro X
+    dsimp
+    simp [eqToHom_map]
+    congr 1
+    · rw [eqToHom_app, eqToHom_trans]
+  map_comp := by
+    intro X Y Z f g
+    dsimp
+    congr 1
+    erw [comp_fiber f g]
+    simp [Functor.comp_map, ← Category.assoc]
+    apply congrFun (congrArg (. ≫ .) ?_) ((α.app Z.base).map g.fiber)
+    slice_rhs 3 3 =>
+      rw [← Functor.comp_map _ _]
+    have H :=  Functor.congr_hom (α.naturality g.base).symm f.fiber
+    erw [H, eqToHom_app, eqToHom_app]
+    simp [Functor.comp_map]
+    erw [eqToHom_app]
+    simp only [Cat.comp_obj, eqToHom_trans, eqToHom_refl, Category.comp_id]
+
+theorem map_obj {α : F ⟶ G} (X : Grothendieck F) :
+    (Grothendieck.map α).obj X = ⟨X.base, (α.app X.base).obj X.fiber⟩ := by
+  rfl
+
+theorem map_map {α : F ⟶ G} {X Y : Grothendieck F} {f : X ⟶ Y} :
+    (Grothendieck.map α).map f =
+    ⟨f.base, (eqToHom (α.naturality f.base).symm).app X.fiber ≫ (α.app Y.base).map f.fiber⟩ := by
+  rfl
+
+/-- The functor `Grothendieck.map α : Grothendieck F ⥤ Grothendieck G` lies over `C`.-/
+theorem functor_comp_forget {α : F ⟶ G} :
+    Grothendieck.map α ⋙ Grothendieck.forget G = Grothendieck.forget F := by
+  rfl
+
+theorem map_id_eq : map (𝟙 F) = 𝟙 (Cat.of <| Grothendieck <| F) := by
+  fapply Functor.ext
+  · intro X
+    rfl
+  · intro X Y f
+    simp [map_map]
+    congr
+    erw [NatTrans.id_app]
+    simp
+
+/-- Making the equality of functors into an isomorphism. Note: we should avoid equality of functors
+if possible, and we should prefer `map_id_iso` to `map_id_eq` whenever we can. -/
+def map_id_iso : map (𝟙 F) ≅ 𝟙 (Cat.of <| Grothendieck <| F) := eqToIso map_id_eq
+
+variable {H : C ⥤ Cat}
+
+theorem map_comp_eq (α : F ⟶ G) (β : G ⟶ H) :
+    map (α ≫ β) = map α ⋙ map β := by
+  fapply Functor.ext
+  · intro X
+    rfl
+  · intro X Y f
+    simp [map_map]
+    fapply Grothendieck.ext
+    · rfl
+    · simp
+      erw [eqToHom_app, eqToHom_app, eqToHom_app, eqToHom_map]
+      simp [eqToHom_trans]
+
+/-- Making the equality of functors into an isomorphism. Note: we should avoid equality of functors
+if possible, and we should prefer `map_comp_iso` to `map_comp_eq` whenever we can. -/
+def map_comp_iso (α : F ⟶ G) (β : G ⟶ H) : map (α ≫ β) ≅ map α ⋙ map β := eqToIso (map_comp_eq α β)
+
+end
+
+universe v
+
+/-- The Grothendieck construction as a functor from the functor category `E ⥤ Cat` to the
+over category `Over E`. -/
+def functor {E : Cat.{v,u}} : (E ⥤ Cat.{v,u}) ⥤ Over (T:= Cat.{v,u}) E where
+  obj F := Over.mk (X:= E) (Y := Cat.of (Grothendieck F))  (Grothendieck.forget F)
+  map {F G} α := Over.homMk (X:= E) (Grothendieck.map α) Grothendieck.functor_comp_forget
+  map_id F := by
+    ext
+    exact Grothendieck.map_id_eq (F:= F)
```

Comment:
Would `map_id_eq` be a bad simp lemma? If it were a simp lemma then automation would take care of this (simliar for `map_comp_eq`)

Reply:
`map_id_eq` and `map_comp_eq` are equalities of functors. I thought we are discouraged from writing `@simp` lemmas for equalities of functors due to their bad rewriting behavior.

Reply:
See for instance in the comment just above docs#CategoryTheory.Functor.comp_id.

Reply:
I see, probably best not to make them simp lemmas then